### ISSUE TITLE
SubAck - Invalid MQTT Control Packet type

### ISSUE THREAD

**Issue by uffebjorklund** [thingsboard/thingsboard#1668]
According to [MQTT Spec](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349805) the SubAck should be 0x90 

When sending a subscription to Thingsboard we get 0x92 back as Control Packet Type. Since this is invalid according to MQTT 3.1.1 our client will close the connection.

Have seen others with the same issue.
https://github.com/micropython/micropython-lib/pull/205

Any plans to fix this? We rather not change the behaviour of the client (breaking the protocol specification)

Regards, Uffe

**Comment by leonmyburg** [comment]
I  belief I have a related issue with umqtt in micropython. No solution yet
https://github.com/thingsboard/thingsboard/issues/1650

**Comment by uffebjorklund** [comment]
Have been looking at various implementations of v3.1.1 and some accept 0x92 (or other SubAcks than 0x90). However after reading the specification a bit more carefully I am convinced that it should be 0x90.

From the specification
```
The remaining bits [3-0] of byte 1 in the fixed header contain flags specific to each MQTT Control Packet type as listed in the Table 2.2 - Flag Bits below. Where a flag bit is marked as “Reserved” in Table 2.2 - Flag Bits, it is reserved for future use and MUST be set to the value listed in that table [MQTT-2.2.2-1]. If invalid flags are received, the receiver MUST close the Network Connection [MQTT-2.2.2-2]. See Section 4.8 for details about handling errors.
```
[See Table 2.2 - Flag Bits](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349758)

**Comment by affoltep** [comment]
@uffebjorklund - Are you able to connect your mqtt client to Thingsboard (TB) using the umqtt micropython lib? I already face a problem to connect due the fact that umqtt supports just username AND password authentication. TB is using the AccessToken as username without any password.

**Comment by Kiogora** [comment]
i have also been affected by this, from wireshark...it identifies header 0x92 as subAck from thingsboard broker but client implementation does not.

**Comment by Kiogora** [comment]
I can confirm the same for pubAck header as 0x42 as opposed to 0x40 according to MQTT 3.1.1 standard.

**Comment by uffebjorklund** [comment]
@Kiogora Seems to be good activity on the project, but still no response on this since I reported it back in April. My best guess is that a PR is needed. We solved our issue by using another solution. Will leave it open for a while since others might have the same problem

**Comment by Kiogora** [comment]
@uffebjorklund What alternative did you use if you dont mind me asking

**Comment by koen-lee** [comment]
This is also a problem when using https://github.com/xamarin/mqtt client or https://github.com/mohaqeq/paho.mqtt.m2mqtt client.

**Comment by yamt** [comment]
MQTT-C (https://github.com/LiamBindle/MQTT-C) complains on this as well. (correctly)

**Comment by Nulifier** [comment]
So I did some looking into this and it seems to stem from these lines:
  * [UnSubAck](https://github.com/thingsboard/thingsboard/blob/b9bc280d57d9a9b84be964fd56aa6f652f23b59d/common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java#L345)
  * [SubAck](https://github.com/thingsboard/thingsboard/blob/b9bc280d57d9a9b84be964fd56aa6f652f23b59d/common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java#L448)
  * [PubAck](https://github.com/thingsboard/thingsboard/blob/b9bc280d57d9a9b84be964fd56aa6f652f23b59d/common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java#L460)

In each of these functions, the fixed MQTT header is created with a QOS of `AT_LEAST_ONCE` where it should be created with `AT_MOST_ONCE` as that represents a 0 value and would be a valid message.

Unfortunately I can't get thingsboard to build for me and I have very little knowledge of Java to figure out why. I created a [branch](https://github.com/Nulifier/thingsboard/tree/fix-mqtt-flags) with the changes I think are required. Is anyone able to build this and verify it fixes the issue?

**Comment by yamt** [comment]
@Nulifier 
we tested your patch. it worked as expected for us.
(our test only covers the SUBACK part of the patch though)

**Comment by trecouvr** [comment]
Opened more than one year ago and still no fix ? We are evaluating going for the PE edition of ThingsBoard but this kind of unsolved issues really hold us back.
We are using the MQTT client from ESP-IDF which is also complaining.

**Comment by AlexDoanTB** [comment]
The fix is available starting v2.5.5 (for AngularJS) and v3.2 (Angular 11)
https://github.com/thingsboard/thingsboard/commit/e3957ea2aa20b8c59435e6d2b435721e930f0510
